### ISSUE TITLE
[Enhancement] Add S3 Client-Side encryption as option

### ISSUE THREAD

**Issue by Belphemur** [nextcloud/server#9304]
Hello,

It would be great to be able to use the Client-Side encryption for AWS instead of having tor rely on NextCloud encryption.

Follow the documentation here: https://docs.aws.amazon.com/aws-sdk-php/v3/guide/service/s3-encryption-client.html

This would make thing so much easier.

**Comment by Belphemur** [comment]
I don't see how it changes the interoperability. I'm not saying the encryption feature should be disabled, or be replaced by S3. Nothing would stop the user to use both encryption if they desire.

What I ask is to add more configuration option into the S3 storage driver, letting the user choose how he want to use the S3 bucket.

This isn't vendor specific as it's using KMS keys that you control.
You'll always be able to decrypt the content/move it to another provider.

For NC, there wouldn't be any change on how it would see the external storage as all the heavy lifting is done by the AWS SDK (and openssl) that is already used by NC to provide the S3 external storage.

In other words, I don't want to replace the current encryption, I just want to be able to activate the one of AWS S3 if I desire or know that it will be the only storage that will be used.

**Comment by KimTheFirst** [comment]
I think this is a no-brainer then - in userland the user should be able to access this feature when managing storage.

I'll bear this in mind myself, as it may be applicable to something I'm doing with another storage provider.

**Comment by ChristophWurst** [comment]
Unfortunately the described feature is not really high on our priority list. But this doesn't mean that we don't like the idea or that we are against this feature per se. Quite the opposite! :slightly_smiling_face: 

There are many way to get new features in Nextcloud:

1. The easiest and most straight forward way (especially here at Github): pull request.
Nextcloud is completely Free Software, this means that everyone is welcome to join and to contribute. If you or anyone else want to work on this feature, this would be great! We appreciate every pull request, and we are definitely able to help in case of questions, reviews, etc.

2. Another option is creating a bounty at [Bountysource](https://www.bountysource.com/search?query=nextcloud), although just putting money on an issue doesn't guarantee that someone picks it up (in time). But in general it is a nice way to support the huge Nextcloud community. In case of a Nextcloud GmbH employee picks up the bounty we will give it back to the community by putting the money back on other bounties to make sure all bounties benefit the Nextcloud community.

3. We have a [category for Freelancers in our form](https://help.nextcloud.com/c/nextcloud-freelancing). Another option would be to post an offer there and try to find a freelancer who want to work on it.

4. The most direct way for a company or organization to get the issue addressed is to get a [Enterprise Subscription](https://nextcloud.com/pricing/). This includes everything to enable you to run Nextcloud in a productive environment with guaranteed SLA's and more. The Enterprise Subscription also includes optional professional services such as custom development. Feel free to [reach out to us](https://nextcloud.com/contact/). We are happy to explore the possibilities how to make Nextcloud fit your needs.
